test/SILOptimizer/utf8_decoding_fastpath.swift

```diff
@@ -12,101 +12,199 @@
+// CHECK-LABEL: end sil function{{.*}}decodeURBPSliceAsUTF8
 public func decodeURBPSliceAsUTF8(_ ptr: Slice<UnsafeBufferPointer<UInt8>>) -> String {
-  blackhole("foo") // otherwise it just jumps into the Slice<UBP> version
   return String(decoding: ptr, as: Unicode.UTF8.self)
 }
+
+public struct CustomContiguousCollection: Collection {
+  let storage: [UInt8]
+  public typealias Index = Int
+  public typealias Element = UInt8
+
+  public init(_ bytes: [UInt8]) { self.storage = bytes }
+  public subscript(position: Int) -> Element { self.storage[position] }
+  public var startIndex: Index { 0 }
+  public var endIndex: Index { storage.count }
+  public func index(after i: Index) -> Index { i+1 }
+
+  @inline(__always)
+  public func withContiguousStorageIfAvailable<R>(
+    _ body: (UnsafeBufferPointer<UInt8>) throws -> R
+  ) rethrows -> R? {
+    try storage.withContiguousStorageIfAvailable(body)
+  }
+}
+public struct CustomNonContiguousCollection: Collection {
+  let storage: [UInt8]
+  public typealias Index = Int
+  public typealias Element = UInt8
+
+  public init(_ bytes: [UInt8]) { self.storage = bytes }
+  public subscript(position: Int) -> Element { self.storage[position] }
+  public var startIndex: Index { 0 }
+  public var endIndex: Index { storage.count }
+  public func index(after i: Index) -> Index { i+1 }
+
+  @inline(__always)
+  public func withContiguousStorageIfAvailable<R>(
+    _ body: (UnsafeBufferPointer<UInt8>) throws -> R
+  ) rethrows -> R? {
+    nil
+  }
+}
+
+// CustomContiguousCollection
+//
+// CHECK-LABEL: sil {{.*}}decodeCustomContiguousAsUTF8{{.*}} : $@convention
+// CHECK-NOT:   function_ref {{.*}}_fromNonContiguousUnsafeBitcastUTF8Repairing
+// CHECK-NOT:   function_ref {{.*}}_fromCodeUnits
+// CHECK:       function_ref {{.*}}_fromUTF8Repairing
+// CHECK-NOT:   function_ref {{.*}}_fromNonContiguousUnsafeBitcastUTF8Repairing
+// CHECK-NOT:   function_ref {{.*}}_fromCodeUnits
+// CHECK-LABEL: end sil function{{.*}}decodeCustomContiguousAsUTF8
+public func decodeCustomContiguousAsUTF8(_ c: CustomContiguousCollection) -> String {
+  return String(decoding: c, as: UTF8.self)
+}
+
+// CustomNonContiguousCollection
+//
+// CHECK-LABEL: sil {{.*}}decodeCustomNonContiguousAsUTF8{{.*}} : $@convention
+// CHECK-NOT:   function_ref {{.*}}_fromCodeUnits
+// CHECK-NOT:   function_ref {{.*}}_fromUTF8Repairing
+// CHECK:       function_ref {{.*}}_fromNonContiguousUnsafeBitcastUTF8Repairing
+// CHECK-NOT:   function_ref {{.*}}_fromCodeUnits
+// CHECK-NOT:   function_ref {{.*}}_fromUTF8Repairing
+// CHECK-LABEL: end sil function{{.*}}decodeCustomNonContiguousAsUTF8
+public func decodeCustomNonContiguousAsUTF8(_ c: CustomNonContiguousCollection) -> String {
+  return String(decoding: c, as: UTF8.self)
+}
+
+// UTF-16
+//
+// NOTE: The SIL optimizer cannot currently fold away a (UTF16.self ==
+// UTF8.self) metatype comparison, so we have to disabel the check-not for UTF-8
+// construction :-(
```

Comment:
@atrick should metatype inequality be foldable as well?

Reply:
It should be easy, just didn't fall out naturally from handling equality. You have to also check that there's no potential relationship between the type values. If you file a bug I or someone else can circle around to it.